arci/src/clients/joint_position_limiter.rs

```diff
@@ -0,0 +1,362 @@
+    pub fn new_with_strategy(
+        client: C,
+        limits: Vec<RangeInclusive<f64>>,
+        strategy: JointPositionLimiterStrategy,
+    ) -> Self {
+        assert!(client.joint_names().len() == limits.len());
+        Self {
+            client,
+            limits,
+            strategy,
+        }
+    }
+
+    pub fn from_urdf(client: C, joints: &[urdf_rs::Joint]) -> Result<Self, Error> {
+        Self::from_urdf_with_strategy(client, joints, Default::default())
+    }
+
+    pub fn from_urdf_with_strategy(
+        client: C,
+        joints: &[urdf_rs::Joint],
+        strategy: JointPositionLimiterStrategy,
+    ) -> Result<Self, Error> {
+        let mut limits = Vec::new();
+        for joint_name in client.joint_names() {
+            if let Some(i) = joints.iter().position(|j| j.name == *joint_name) {
+                let joint = &joints[i];
+                let limit = if JointType::Continuous == joint.joint_type {
+                    // If limit is not specified, urdf-rs assigns f64::default.
+                    -f64::consts::PI..=f64::consts::PI
+                } else {
+                    joint.limit.lower..=joint.limit.upper
+                };
+                limits.push(limit);
+            } else {
+                return Err(Error::NoJoint(joint_name.into()));
+            }
+        }
+
+        Ok(Self {
+            client,
+            limits,
+            strategy,
+        })
+    }
+
+    pub fn set_strategy(&mut self, strategy: JointPositionLimiterStrategy) {
+        self.strategy = strategy;
+    }
+
+    fn check_joint_position(&self, positions: &mut Vec<f64>) -> Result<(), Error> {
+        assert_eq!(positions.len(), self.limits.len());
+        for (i, (limit, position)) in self.limits.iter().zip(positions).enumerate() {
+            if limit.contains(&position) {
+                continue;
+            }
+            match self.strategy {
+                JointPositionLimiterStrategy::Error => {
+                    return Err(Error::OutOfLimit {
+                        name: self.client.joint_names()[i].clone(),
+                        position: *position,
+                        limit: limit.clone(),
+                    });
+                }
+                JointPositionLimiterStrategy::Round => {
+                    debug!(
+                        "Out of limit: joint={}, position={}, limit={:?}",
+                        self.client.joint_names()[i],
+                        position,
+                        limit,
+                    );
+                    *position = position.clamp(*limit.start(), *limit.end());
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+impl<C> JointTrajectoryClient for JointPositionLimiter<C>
+where
+    C: JointTrajectoryClient,
+{
+    fn joint_names(&self) -> &[String] {
+        self.client.joint_names()
+    }
+
+    fn current_joint_positions(&self) -> Result<Vec<f64>, Error> {
+        self.client.current_joint_positions()
```

Comment:
This may be optional, but it might be a good idea to use `check_joint_position()` here.
In the real world, the robot returns the values out of the limits, and the system can't handle the values which are out of the limit. If this handles this, it reduces the cost to check the current_joint_positions in the user logics.

Reply:
done in 1a5d5e1